### ISSUE TITLE
How to use dataflow analysis of MIR to track the variable's value

### ISSUE THREAD

**Issue by relaxcn** [rust-lang/rust-clippy#14353]
I'm writting a new lint.

I want to use **rustc_mir_dataflow** to track the variable's value.

For example:

```rust
fn fun(num: String) -> String {
    num
}

fn foo() {
    let s = String::from("hello");
    let s1 = s;
    let s2 = fun(s1);
}
```

I just want to know, which variable is assigned the value of `s` , and I want to track where the data of this variable is passed to.

I know I need to use MIR and CFG for analysis, but I have no experience about it.

Does anyone give me some suggestions? What I should to do or what I need to learn?

I searched the lints about MIR in rust-clippy source repo, but I can only find few lints about it.

Thank you for you help.

**Comment by relaxcn** [comment]
And I know there is a helpful struct `PossibleBorrowerMap` in `clippy_utils::mir`. It's usefull.

But I am confused about how to convert `mir::Local` to `hir::hir_id`.

**Comment by smoelius** [comment]
> But I am confused about how to convert `mir::Local` to `hir::hir_id`.

This is not what you asked for, but this function tries to determine the locals that correspond to an expression:
https://github.com/rust-lang/rust-clippy/blob/2440f516965ea948e9d04e56c63beee36deb8848/clippy_utils/src/mir/mod.rs#L149

If what you want doesn't exist, that function might serve as a starting point.

**Comment by relaxcn** [comment]
Thanks for your help.
I just do not want to use `BorrowerMap`, I want to track value passing including by function call.
Do you have any ideas?

**Comment by smoelius** [comment]
> Thanks for your help. I just do not want to use `BorrowerMap`, I want to track value passing including by function call. Do you have any ideas?

I think the MIR dataflow framework is meant for analyzing single function/closure bodies. So if you're trying to doing an inter-procedural analysis, that could be tricky.

But here is a [useful tidbit](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.local_decls) if you're trying to track values passed to a function:
> The first local is the return value pointer, followed by arg_count locals for the function arguments, followed by any user-declared variables and temporaries.

You've probably found this, but the dataflow section of the Rust Compiler Developer Guide is very helpful: https://rustc-dev-guide.rust-lang.org/mir/dataflow.html#dataflow-analysis

Also, you might consider searching the Rust repo for examples of the dataflow framework's use. For example, here is one: https://github.com/rust-lang/rust/blob/4559163ccb500affc424fb9228dae5003672ffc7/compiler/rustc_mir_dataflow/src/impls/storage_liveness.rs#L27

I hope the points are helpful.

**Comment by relaxcn** [comment]
Thanks a lot!
I have learned how to do the dataflow analysis.
I'm trying to get more information about it and trying some simple examples.
I have a problem:

This is my use case:

```rust
fn main() {
    let a = 1;
}
```

This is my lint:

```rust
impl LateLintPass<'_> for DataflowTest {
    fn check_fn(&mut self, cx: &LateContext<'_>,_:rustc_hir::intravisit::FnKind<'_>,_: &'_ rustc_hir::FnDecl<'_>, _: &'_ rustc_hir::Body<'_>,_:rustc_span::Span, def_id :rustc_span::def_id::LocalDefId) {
        dbg!("fn");
        let tcx = cx.tcx;
        let body = cx.tcx.optimized_mir(def_id);
        let storage = MyStorage{};
        let mut results = storage.iterate_to_fixpoint(tcx, body, None).into_results_cursor(body);
    }
}

struct MyStorage{}

impl<'tcx> Analysis<'tcx> for MyStorage {
    type Domain = DenseBitSet<Local>;

    const NAME: &'static str = "mystorage";

    fn bottom_value(&self, body: &Body<'tcx>) -> Self::Domain {
        eprintln!("Calling bottom_value for body with {} blocks", body.basic_blocks.len());
        // bottom = dead
        DenseBitSet::new_empty(body.local_decls.len())

    }

    fn initialize_start_block(&self, body: &Body<'tcx>, state: &mut Self::Domain) {
    }

    fn apply_primary_statement_effect(
        &mut self,
        state: &mut Self::Domain,
        stmt: &Statement<'tcx>,
        _: Location,
    ) {
    }
}
```

I don't know why the string "Calling bottom_value for body with 1 blocks" is printed three times?

```
full stderr:
[clippy_lints/src/dataflow_test.rs:32:9] "fn" = "fn"
Calling bottom_value for body with 1 blocks
Calling bottom_value for body with 1 blocks
Calling bottom_value for body with 1 blocks
```

I think the use case has just one block, so the function `bottom_value` should be excuted once.

**Comment by Jarcho** [comment]
Two extra is the right amount. Both `iterate_to_fixpoint` and the cursor have an extra intermediary state initialized with `bottom_value`. Both are needed since the current block entry states have to be kept while applying the statement and terminator effects.
